lib/internal/wasi.js

```diff
@@ -0,0 +1,1301 @@
+            }
+            read += r;
+          });
+        this.view.setUint32(nread, read, true);
+        return WASI_ESUCCESS;
+      }),
+      fd_read: wrap((fd, iovs, iovsLen, nread) => {
+        const stats = CHECK_FD(fd, WASI_RIGHT_FD_READ);
+        let read = 0;
+        outer:
+        for (const iov of getiovs(iovs, iovsLen)) {
+          let r = 0;
+          while (r < iov.byteLength) {
+            const rr = fs.readSync(stats.real, iov, r, iov.byteLength - r);
+            r += rr;
+            read += rr;
+            if (rr === 0) {
+              break outer;
+            }
+          }
+        }
+        this.view.setUint32(nread, read, true);
+        return WASI_ESUCCESS;
+      }),
+      fd_readdir: wrap((fd, bufPtr, bufLen, cookie, bufusedPtr) => {
+        const stats = CHECK_FD(fd, WASI_RIGHT_FD_READDIR);
+        this.refreshMemory();
+        const entries = fs.readdirSync(stats.path, { withFileTypes: true });
+        const startPtr = bufPtr;
+        for (let i = Number(cookie); i < entries.length; i += 1) {
+          const entry = entries[i];
+          this.view.setBigUint64(bufPtr, BigInt(i + 1), true);
+          bufPtr += 8;
+          const rstats = fs.statSync(path.resolve(stats.path, entry.name));
+          this.view.setBigUint64(bufPtr, BigInt(rstats.ino), true);
+          bufPtr += 8;
+          this.view.setUint32(bufPtr, Buffer.byteLength(entry.name), true);
+          bufPtr += 4;
+          let filetype;
+          switch (true) {
+            case rstats.isBlockDevice():
+              filetype = WASI_FILETYPE_BLOCK_DEVICE;
+              break;
+            case rstats.isCharacterDevice():
+              filetype = WASI_FILETYPE_CHARACTER_DEVICE;
+              break;
+            case rstats.isDirectory():
+              filetype = WASI_FILETYPE_DIRECTORY;
+              break;
+            case rstats.isFIFO():
+              filetype = WASI_FILETYPE_SOCKET_STREAM;
+              break;
+            case rstats.isFile():
+              filetype = WASI_FILETYPE_REGULAR_FILE;
+              break;
+            case rstats.isSocket():
+              filetype = WASI_FILETYPE_SOCKET_STREAM;
+              break;
+            case rstats.isSymbolicLink():
+              filetype = WASI_FILETYPE_SYMBOLIC_LINK;
+              break;
+            default:
+              filetype = WASI_FILETYPE_UNKNOWN;
+              break;
+          }
+          this.view.setUint8(bufPtr, filetype);
+          bufPtr += 1;
+          bufPtr += 3; // padding
+          Buffer.from(this.memory.buffer)
+            .write(entry.name, bufPtr, bufLen - bufPtr);
+          bufPtr += Buffer.byteLength(entry.name);
+          bufPtr += (8 % bufPtr); // padding
+        }
+        const bufused = bufPtr - startPtr;
+        this.view.setUint32(bufusedPtr, bufused, true);
+        return WASI_ESUCCESS;
+      }),
+      fd_renumber: wrap((from, to) => {
+        CHECK_FD(from, 0);
+        CHECK_FD(to, 0);
```

Comment:
Does `to` have to be an open file descriptor? (I haven’t checked, but as this is supposed to be similar to `dup2()`, that would make sense to me)

Reply:
CHECK_FD should fail if the fd isn't already in the fd map

Reply:
@devsnek Yes, that’s what I’m worried about – is `fd_renumber` invalid if the to-be-overwritten fd is nonexistent? That would seem a bit odd to me, although it’s totally possible that that’s the case